.github/workflows/docker.yml

```diff
@@ -0,0 +1,166 @@
+        set -e
+        TAG=$( git describe --tags --dirty ) # E.g. v1.2.0-23-g60ee190
+        echo "TAG=$TAG" >> $GITHUB_ENV
+
+    # See: https://github.com/docker/build-push-action/blob/v2.6.1/docs/advanced/cache.md#github-cache
+    - name: Set up QEMU
+      uses: docker/setup-qemu-action@v2
+
+    - name: Set up Docker Buildx
+      id: buildx
+      uses: docker/setup-buildx-action@master
+
+    - name: Cache Docker layers
+      uses: actions/cache@v3
+      with:
+        path: /tmp/.buildx-cache
+        key: ${{ runner.os }}-buildx-${{ github.sha }}
+        restore-keys: |
+          ${{ runner.os }}-buildx-
+
+    - name: Login to Docker Hub registry
+      uses: docker/login-action@v2
+      with:
+        username: ${{ secrets.DOCKERHUB_USERNAME }}
+        password: ${{ secrets.DOCKERHUB_TOKEN }}
+
+    - name: 'Login to GitHub Container Registry'
+      uses: docker/login-action@v2
+      with:
+        registry: ghcr.io
+        username: ${{ github.actor }}
+        password: ${{ secrets.GITHUB_TOKEN }}
+
+    - uses: actions/download-artifact@v3
+      with:
+        name: binaries-${{ github.sha }}
+
+    - run: |
+        ls -al release*/
+
+    - name: Build (PRs)
+      id: docker_build_pr
+      # Run only on pull requests
+      if: github.event_name == 'pull_request'
+      uses: docker/build-push-action@v3
+      with:
+        build-args: |
+          TAG=${{ env.TAG }}
+        context: '.'
+        file: Dockerfile
+        platforms: linux/amd64,linux/arm,linux/arm64,linux/mips64le,linux/ppc64le,linux/riscv64,linux/s390x
+        push: true
+        tags: |
+          ${{ github.repository }}:${{ env.TAG }}
+          ghcr.io/${{ github.repository }}:${{ env.TAG }}
+        cache-from: type=local,src=/tmp/.buildx-cache
+        cache-to: type=local,dest=/tmp/.buildx-cache-new,mode=max
+
+    - name: Build and push (master)
+      id: docker_build_master
+      # Run only on master
+      if: github.ref == 'refs/heads/master'
+      uses: docker/build-push-action@v3
+      with:
+        build-args: |
+          TAG=${{ env.TAG }}
+        context: '.'
+        file: Dockerfile
+        platforms: linux/amd64,linux/arm,linux/arm64,linux/mips64le,linux/ppc64le,linux/riscv64,linux/s390x
+        push: true
+        tags: |
+          ${{ github.repository }}:${{ env.TAG }}
+          ghcr.io/${{ github.repository }}:${{ env.TAG }}
+        cache-from: type=local,src=/tmp/.buildx-cache
+        cache-to: type=local,dest=/tmp/.buildx-cache-new,mode=max
+
+    - name: Build and push (release)
+      id: docker_build_release
+      if: startsWith(github.ref, 'refs/tags/')
+      uses: docker/build-push-action@v3
+      with:
+        build-args: |
+          TAG=${{ env.TAG }}
+        context: '.'
+        file: Dockerfile
+        platforms: linux/amd64,linux/arm,linux/arm64,linux/mips64le,linux/ppc64le,linux/riscv64,linux/s390x
+        push: true
+        tags: |
+          ${{ github.repository }}:${{ env.TAG }}
```

Comment:
This is a total extra credit, but it would be cool if we could push both, say, `v1.2.3`, `v1.2`, and `v1` tags here. So that people don't have to rely on `latest`.

Reply:
yes.

Reply:
done.